### ISSUE TITLE
`ForwardRef` is broken on 3.12

### ISSUE THREAD

**Issue by PerchunPak** [alexmojaki/eval_type_backport#24]
3.12 added a new argument, which your wrapper doesn't accept and therefore crashes.

https://github.com/alexmojaki/eval_type_backport/actions/runs/11103634046/job/30846002124#step:5:116
https://github.com/python/cpython/commit/5430f614371530aab1178b3b610add4bf54c383e#diff-ddb987fca5f5df0c9a2f5521ed687919d70bb3d64eaeb8021f98833a2a716887R916

**Comment by johnslavik** [comment]
This library is useless in Python versions higher than 3.10.
I recommend using a snippet like
```py
if sys.version_info >= (3, 11):
    from typing import ForwardRef
else:
    from eval_type_backport import ForwardRef
```
to work around the breaking change.

**Comment by PerchunPak** [comment]
Well, I am a package maintainer for this library on NixOS and a few packages have tests failures after updating from 0.1.3 to 0.2.0

- [immich-machine-learning](https://github.com/immich-app/immich/tree/main/machine-learning)
- [python312Packages.albumentations](https://github.com/albumentations-team/albumentations)
- [python312Packages.insightface](https://github.com/deepinsight/insightface)
- [unstructured-api](https://github.com/Unstructured-IO/unstructured-api)

Changing code of those packages is obviously not an option, would it be okay if I will submit a PR fixing this?

**Comment by DanCardin** [comment]
Seems like these packages likely are depending on this one transitively through e.g. pydantic, and (to me it seems like) this is more of a bug in how pydantic is calling into this library in python versions where it shouldnt (at least at a glance, i dont see that it is making conditional use of this library based on python version).

**Comment by PerchunPak** [comment]
Here is a dependency tree for those packages

- [immich-machine-learning](https://github.com/NixOS/nixpkgs/blob/d3c42f187194c26d9f0309a8ecc469d6c878ce33/pkgs/by-name/im/immich-machine-learning/package.nix#L33)
  - [insightface](https://github.com/NixOS/nixpkgs/blob/d3c42f187194c26d9f0309a8ecc469d6c878ce33/pkgs/development/python-modules/insightface/default.nix#L46)
    - [albumentations](https://github.com/NixOS/nixpkgs/blob/d3c42f187194c26d9f0309a8ecc469d6c878ce33/pkgs/development/python-modules/albumentations/default.nix#L56)
      - eval-type-backport

No pydantic involved. In fact, eval-type-backport is installed with pydantic [only if Python version is 3.10 or lower](https://github.com/NixOS/nixpkgs/blob/d3c42f187194c26d9f0309a8ecc469d6c878ce33/pkgs/development/python-modules/pydantic/default.nix#L71).

**Comment by alexmojaki** [comment]
@DanCardin here's how the import in pydantic depends on Python version:

https://github.com/pydantic/pydantic/blob/5bd3a6507b749fcd4833173fba88b3690ff77170/pydantic/_internal/_typing_extra.py#L635-L640

https://github.com/pydantic/pydantic/blob/5bd3a6507b749fcd4833173fba88b3690ff77170/pydantic/_internal/_typing_extra.py#L672-L680

@PerchunPak I don't see any mentions of eval-type-backport in the code of these libraries, so they're almost certainly just using it via pydantic. I'm curious where tests are failing in those libraries. But I agree that this should be fixed here anyway, and a PR would be appreciated.

**Comment by alexmojaki** [comment]
BTW in Python 3.13 `typing._eval_type` also accepts a new `type_params` argument, so that should be added to the `eval_type_backport` function.

**Comment by DanCardin** [comment]
> @DanCardin here's how the import in pydantic depends on Python version:

idk if i'm being crazy here but...

```python
def is_backport_fixable_error(e: TypeError) -> bool:
    msg = str(e)

    return (
        sys.version_info < (3, 10)  # in 3.12, false
        and msg.startswith('unsupported operand type(s) for |: ')
        or sys.version_info < (3, 9)  # in 3.12, false
        and "' object is not subscriptable" in msg
    )  # thus this whole thing is false

...
        if (
           not (isinstance(value, typing.ForwardRef)  # presumably can be true or false
           and is_backport_fixable_error(e)  # thus false
       ):
            raise  # thus this isn't hit

       # thus it calls eval_type_backport
```

to me, this reads as that it will call to this library in 3.12, unless i'm missing something fundamental. And i guess clearly it is being called in 3.12, given the bug report.

While it might be something this library wants to address anyways, it still seems like pydantic should have a much clearer: if the version is > x, dont call it ever, condition.

**Comment by alexmojaki** [comment]
@DanCardin you're reading the parentheses wrong in the second part

**Comment by PerchunPak** [comment]
It is

```py
if not (
    isinstance(value, typing.ForwardRef)  # presumably can be true or false
    and is_backport_fixable_error(e)  # false (see above)
):  # thus true

**Comment by alexmojaki** [comment]
Released the fix now in version 0.2.1
